Intermittent failure in TGMathTest.trigonometrics on Apple M1 hardware.

Since I am prone to ranting when the work I an trying to do gets interrupted, let
me try to describe what I am experiencing with some kindness in my heart and
on my keyboard.

##### Presenting problem

I get intermittent failures when when running the TGMathTest class on macOs (Monterrey 12.3 or such) on M1 hardware.
Sometimes the failures are as often as every other run. Others they are 1 in 5 or so.  I have never seen this
problem in CI or on my personal Linux machines.
 
```
[error] Test scala.scalanative.libc.TGMathTest.trigonometrics failed: java.lang.AssertionError: expected:<1.9397751911942611> but was:<1.9397751911942618>, took 0.001 sec
```

##### Discussion
I hate Heisenbugs! I would appreciate any discussion and/or opinions
on the issue. I know "Glad it is you and not me" is one of those thoughts.

I asked on Discord and I think that other people have seen this problem.

There are at lest two systematic problems in the TGMathTest class
1) There tend to be more than one assertion in a Test and those assertions have no distinguishing message.
2) In a number of places the values tested are randomly generated.

Problem 1 makes it very hard to determine _which_ assertion failed. This disturbs
harmony.

Problem 2 makes it __very_hard__ to reproduce problems with specific values. This leads
to the experienced indeterminacy. There is at least one long discussion about whether
random values should be used in unit-tests to spread coverage or if all values should
be fixed.  I will mumble _soto_voce_ and bypass that discussion.

From the first time I experienced this problem I believe there is some issue
that some of the Random values can get very close to 0 or very close to 1.
If I remember my long ago Numerical Methods classes, math with such
numbers is tricky; trig with such numbers is even trickier. I believe that
most numerical libraries use table lookup for some/many of those numbers.
Again, from that long ago debugging week, I seem to remember that
the underlying problem here is that a dodgy some (sin(x) / cos) explicit
math is done and then compared to a library lookup.  In the given
time I could not track down the offending values. I suspect that 
x was either near 0 or 1.  The calculated and the library values
were compared and expected to be within some tiny epsilon 
(1 or 2) ulp (unit of least precision). I think the difference
as actually much greater than that.

Someone who uses double precision trig functions on
a daily basis may have a better answer. It looks to me
like the test if validly picking up some questionable
numerical accuracy in the test itself, as implemented.

I think fixing this problem is going to take a number of
iterations.  I believe that it would be a useful first pass
to have give a message to each of the assertions
in the `trigonomics` test.  Yes, in a better world, 
one would do every assert in the file. For now, let's
at least find the problem which is breaking private M1 builds.

I have several pages of debugging notes which I need study.  The obvious result
so far is that the calculated value, scala jvm, scala native, python, etc. seems to
agree out to 16 or so places.  Scala  and python library calls to `tan()` appear to
match, but Scala Native's M1 library call results are the odd person out.  I need
to dive into the SN implementation. Scala JVM, Python are almost
certainly calling the same system library (libc?). Does the M1 library differ
all that much?

The information below isolates the failing code and gives a few values
which cause failure. I hope it may be useful to someone while I debug
this some more.

```
Code:

      // Leads to 0.0 (inclusive) to 100.0 (exclusive) radians.
      // That is a _lot_ of radians!  Anything over 2 pi is bragging.

      val ns = List.fill(100)(scala.util.Random.nextDouble() * 100)
      ns.foreach { n =>
      ... 
      val tanN0 = tgmath.sin(n) / tgmath.cos(n)
      val tanN1 = tgmath.tan(n)
      val tanNf0 = tgmath.sin(n.toFloat) / tgmath.cos(n.toFloat)
      val tanNf1 = tgmath.tan(n.toFloat)

      assertEquals(
	s"n: ${n} tanN0: ${tanN0} tanN1: ${tanN1}",
        tanN0,
        tanN1,
        TestUtils.eps(tanN0, tanN1)
      )

[error] Test scala.scalanative.libc.TGMathTest.trigonometrics failed: java.lang\
.AssertionError: n: 13.281526950995149 tanN0: 0.8685344429479043 tanN1: 0.86853\
44429479046 expected:<0.8685344429479043> but was:<0.8685344429479046>, took 0.\
002 sec
[error] Test scala.scalanative.libc.TGMathTest.trigonometrics failed: java.lang\
.AssertionError: n: 63.83967211742178 tanN0: 1.584522756925413 tanN1: 1.5845227\
569254123 expected:<1.584522756925413> but was:<1.5845227569254123>, took 0.001\
 sec

AssertionError: n: 79.27422084268272 tanN0: 0.9028809395817675 tanN1: 0.9028809\
395817672 expected:<0.9028809395817675> but was:<0.9028809395817672>, took 0.00\
2 sec

For grins & giggles, I commented out the offending tanN0 assertion, leaving
the assertion:
      assertEquals(
        s"n: ${n} tanNf0: ${tanNf0} tanNf1: ${tanNf1}",
        tanNf0,
        tanNf1,
        TestUtils.eps(tanNf0, tanNf1)
      )

That ran 20 times without failure. So, it is the digits beyond a float
which are causing the problem: tiny values, probably not contiguous
or within 1 ulp.
```

--- a/unit-tests/native/src/test/scala/scala/scalanative/TestUtils.scala
+++ b/unit-tests/native/src/test/scala/scala/scalanative/TestUtils.scala
@@ -9,8 +9,19 @@ private[scalanative] object TestUtils {
     val diff = Math.abs(act - exp)
     diff <= eps(act, exp)
   }
+
+  /* The mathematically inclined and informed will notice that 'eps' here
+   * is a misnomer & misdirection (a.k.a lie).
+   *
+   * These methods have the proper math meaning only when nUlp == 1.
+   */
+
+  def eps(actual: Double, expected: Double, nUlp: Int): Double =
+    nUlp * Math.max(Math.ulp(actual), Math.ulp(expected))
+
   def eps(actual: Double, expected: Double): Double =
     2 * Math.max(Math.ulp(actual), Math.ulp(expected))
+
   def eps(actual: Float, expected: Float): Float =
     2 * Math.max(Math.ulp(actual), Math.ulp(expected))
 }

--- a/unit-tests/native/src/test/scala/scala/scalanative/libc/TGMathTest.scala
+++ b/unit-tests/native/src/test/scala/scala/scalanative/libc/TGMathTest.scala
@@ -6,6 +6,9 @@ import org.junit.Assert._
 
 import scalanative.unsafe._
 import scalanative.libc._
+import scalanative.runtime.Platform
+import scalanative.runtime.PlatformExt
+
 class TGMathTest {
 
   // basic math stuffs
@@ -290,6 +293,7 @@ class TGMathTest {
       TestUtils.eps(pi, PI)
     )
   }
+
   // -1 <= sin theta <=1, -1 <= cos theta <= 1, sin theta^2 + cos theta^2 = 1
   @Test def trigonometrics(): Unit = {
     val ns = List.fill(100)(scala.util.Random.nextDouble() * 100)
@@ -298,6 +302,7 @@ class TGMathTest {
       assertTrue(tgmath.sin(n.toFloat) >= -1.0)
       assertTrue(tgmath.sin(n) <= 1.0)
       assertTrue(tgmath.sin(n.toFloat) <= 1.0)
+
       val sin2_cos2 =
         tgmath.pow(tgmath.sin(n), 2.0) + tgmath.pow(tgmath.cos(n), 2.0)
       assertEquals(
@@ -305,15 +310,36 @@ class TGMathTest {
         sin2_cos2,
         TestUtils.eps(1.0, sin2_cos2)
       )
+
       val tanN0 = tgmath.sin(n) / tgmath.cos(n)
       val tanN1 = tgmath.tan(n)
+
       val tanNf0 = tgmath.sin(n.toFloat) / tgmath.cos(n.toFloat)
       val tanNf1 = tgmath.tan(n.toFloat)
+
+      /* Scala Native Issue #2641
+       *
+       * The tan(double) function on Apple M1 is known, to some and
+       * circa June 2022, as having poor numerical accuracy (> 3 eps).
+       * tan(float) is accurate. See:
+       *   https://members.loria.fr/PZimmermann/papers/accuracy.pdf
+       *
+       * Use wider acceptable bounds for Apple M1 tan(double).
+       */
+      // PlatformExt.isArm64 is currently broken/unreliable.
+      // isMac() test here weakens the bounds on both Intel and Arm64
+      // (i.e. M1). macOS on Intel hardware is known to work within 2 eps.
+      //
+      // The condition when a  truthful isArm64 is available should be:
+      // (Platform.isMac() && PlatformExt.isArm64)
+      val nEps = if (Platform.isMac()) 4 else 2
+
       assertEquals(
         tanN0,
         tanN1,
-        TestUtils.eps(tanN0, tanN1)
+        TestUtils.eps(tanN0, tanN1, nEps)
       )
+
       assertEquals(
         tanNf0,
         tanNf1,
@@ -321,6 +347,7 @@ class TGMathTest {
       )
     }
   }
+
   @Test def hyperbolics(): Unit = {
     val ns = List.fill(100)(scala.util.Random.nextDouble() * 100)
     ns.foreach { n =>

